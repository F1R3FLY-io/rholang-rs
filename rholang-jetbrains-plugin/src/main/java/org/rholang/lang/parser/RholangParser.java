package org.rholang.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

public class RholangParser implements PsiParser {
    private static final Logger LOG = Logger.getInstance(RholangParser.class);

    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker rootMarker = builder.mark();

        // Get the full text from the builder
        String text = builder.getOriginalText().toString();

        // Check if the code is valid using the Rust parser
        boolean isValid = RholangParserCli.isValid(text);
        LOG.debug("Rholang code is valid: " + isValid);

        // For now, just consume all tokens
        // In a more complete implementation, we would parse the tree returned by the Rust parser
        // and build a corresponding PSI tree
        while (!builder.eof()) {
            builder.advanceLexer();
        }

        rootMarker.done(root);
        return builder.getTreeBuilt();
    }
}
