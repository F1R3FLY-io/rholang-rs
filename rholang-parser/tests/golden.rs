use rholang_parser::RholangParser;
use rstest::rstest;
use std::fs;
use std::path::PathBuf;

#[rstest]
fn golden_test(
    #[base_dir = "tests/corpus/"]
    #[files("*.rho")]
    path: PathBuf,
) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("corpus/golden_snapshots");

    settings.bind(|| {
        let name = path.file_stem().unwrap().to_string_lossy().to_string();

        let parser = RholangParser::new();
        let input = fs::read_to_string(path).expect("Failed to read input file");
        let result = parser.parse(input.as_str());

        // Each snapshot is named after the input file
        insta::assert_debug_snapshot!(name, result);
    })
}
