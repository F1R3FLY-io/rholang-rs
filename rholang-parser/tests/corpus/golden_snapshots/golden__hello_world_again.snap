---
source: rholang-parser/tests/golden.rs
expression: result
---
Good(
    [
        AnnProc {
            proc: New {
                decls: [
                    NameDecl {
                        id: Id {
                            name: "helloworld",
                            pos: SourcePos {
                                line: 3,
                                col: 5,
                            },
                        },
                        uri: None,
                    },
                    NameDecl {
                        id: Id {
                            name: "stdout",
                            pos: SourcePos {
                                line: 3,
                                col: 17,
                            },
                        },
                        uri: Some(
                            Uri(
                                "rho:io:stdout",
                            ),
                        ),
                    },
                ],
                proc: AnnProc {
                    proc: Par {
                        left: AnnProc {
                            proc: Contract {
                                name: NameVar(
                                    Id(
                                        Id {
                                            name: "helloworld",
                                            pos: SourcePos {
                                                line: 4,
                                                col: 14,
                                            },
                                        },
                                    ),
                                ),
                                formals: Names {
                                    names: [
                                        NameVar(
                                            Id(
                                                Id {
                                                    name: "world",
                                                    pos: SourcePos {
                                                        line: 4,
                                                        col: 26,
                                                    },
                                                },
                                            ),
                                        ),
                                    ],
                                    remainder: None,
                                },
                                body: AnnProc {
                                    proc: ForComprehension {
                                        receipts: [
                                            [
                                                Linear {
                                                    lhs: Names {
                                                        names: [
                                                            Quote(
                                                                AnnProc {
                                                                    proc: ProcVar(
                                                                        Id(
                                                                            Id {
                                                                                name: "msg",
                                                                                pos: SourcePos {
                                                                                    line: 5,
                                                                                    col: 15,
                                                                                },
                                                                            },
                                                                        ),
                                                                    ),
                                                                    span: SourceSpan {
                                                                        start: SourcePos {
                                                                            line: 5,
                                                                            col: 15,
                                                                        },
                                                                        end: SourcePos {
                                                                            line: 5,
                                                                            col: 18,
                                                                        },
                                                                    },
                                                                },
                                                            ),
                                                        ],
                                                        remainder: None,
                                                    },
                                                    rhs: Simple {
                                                        name: NameVar(
                                                            Id(
                                                                Id {
                                                                    name: "world",
                                                                    pos: SourcePos {
                                                                        line: 5,
                                                                        col: 22,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                    },
                                                },
                                            ],
                                        ],
                                        proc: AnnProc {
                                            proc: Send {
                                                channel: NameVar(
                                                    Id(
                                                        Id {
                                                            name: "stdout",
                                                            pos: SourcePos {
                                                                line: 6,
                                                                col: 13,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                send_type: Single,
                                                inputs: [
                                                    AnnProc {
                                                        proc: ProcVar(
                                                            Id(
                                                                Id {
                                                                    name: "msg",
                                                                    pos: SourcePos {
                                                                        line: 6,
                                                                        col: 21,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 6,
                                                                col: 21,
                                                            },
                                                            end: SourcePos {
                                                                line: 6,
                                                                col: 24,
                                                            },
                                                        },
                                                    },
                                                ],
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 6,
                                                    col: 13,
                                                },
                                                end: SourcePos {
                                                    line: 6,
                                                    col: 25,
                                                },
                                            },
                                        },
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 5,
                                            col: 9,
                                        },
                                        end: SourcePos {
                                            line: 7,
                                            col: 10,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 4,
                                    col: 5,
                                },
                                end: SourcePos {
                                    line: 8,
                                    col: 6,
                                },
                            },
                        },
                        right: AnnProc {
                            proc: New {
                                decls: [
                                    NameDecl {
                                        id: Id {
                                            name: "world",
                                            pos: SourcePos {
                                                line: 9,
                                                col: 9,
                                            },
                                        },
                                        uri: None,
                                    },
                                    NameDecl {
                                        id: Id {
                                            name: "world2",
                                            pos: SourcePos {
                                                line: 9,
                                                col: 16,
                                            },
                                        },
                                        uri: None,
                                    },
                                ],
                                proc: AnnProc {
                                    proc: Par {
                                        left: AnnProc {
                                            proc: Par {
                                                left: AnnProc {
                                                    proc: Par {
                                                        left: AnnProc {
                                                            proc: Send {
                                                                channel: NameVar(
                                                                    Id(
                                                                        Id {
                                                                            name: "helloworld",
                                                                            pos: SourcePos {
                                                                                line: 10,
                                                                                col: 9,
                                                                            },
                                                                        },
                                                                    ),
                                                                ),
                                                                send_type: Single,
                                                                inputs: [
                                                                    AnnProc {
                                                                        proc: Eval {
                                                                            name: NameVar(
                                                                                Id(
                                                                                    Id {
                                                                                        name: "world",
                                                                                        pos: SourcePos {
                                                                                            line: 10,
                                                                                            col: 22,
                                                                                        },
                                                                                    },
                                                                                ),
                                                                            ),
                                                                        },
                                                                        span: SourceSpan {
                                                                            start: SourcePos {
                                                                                line: 10,
                                                                                col: 21,
                                                                            },
                                                                            end: SourcePos {
                                                                                line: 10,
                                                                                col: 27,
                                                                            },
                                                                        },
                                                                    },
                                                                ],
                                                            },
                                                            span: SourceSpan {
                                                                start: SourcePos {
                                                                    line: 10,
                                                                    col: 9,
                                                                },
                                                                end: SourcePos {
                                                                    line: 10,
                                                                    col: 28,
                                                                },
                                                            },
                                                        },
                                                        right: AnnProc {
                                                            proc: Send {
                                                                channel: NameVar(
                                                                    Id(
                                                                        Id {
                                                                            name: "world",
                                                                            pos: SourcePos {
                                                                                line: 11,
                                                                                col: 9,
                                                                            },
                                                                        },
                                                                    ),
                                                                ),
                                                                send_type: Single,
                                                                inputs: [
                                                                    AnnProc {
                                                                        proc: StringLiteral(
                                                                            "Hello World",
                                                                        ),
                                                                        span: SourceSpan {
                                                                            start: SourcePos {
                                                                                line: 11,
                                                                                col: 16,
                                                                            },
                                                                            end: SourcePos {
                                                                                line: 11,
                                                                                col: 29,
                                                                            },
                                                                        },
                                                                    },
                                                                ],
                                                            },
                                                            span: SourceSpan {
                                                                start: SourcePos {
                                                                    line: 11,
                                                                    col: 9,
                                                                },
                                                                end: SourcePos {
                                                                    line: 11,
                                                                    col: 30,
                                                                },
                                                            },
                                                        },
                                                    },
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 10,
                                                            col: 9,
                                                        },
                                                        end: SourcePos {
                                                            line: 11,
                                                            col: 30,
                                                        },
                                                    },
                                                },
                                                right: AnnProc {
                                                    proc: Send {
                                                        channel: NameVar(
                                                            Id(
                                                                Id {
                                                                    name: "helloworld",
                                                                    pos: SourcePos {
                                                                        line: 12,
                                                                        col: 9,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        send_type: Single,
                                                        inputs: [
                                                            AnnProc {
                                                                proc: Eval {
                                                                    name: NameVar(
                                                                        Id(
                                                                            Id {
                                                                                name: "world2",
                                                                                pos: SourcePos {
                                                                                    line: 12,
                                                                                    col: 22,
                                                                                },
                                                                            },
                                                                        ),
                                                                    ),
                                                                },
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 12,
                                                                        col: 21,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 12,
                                                                        col: 28,
                                                                    },
                                                                },
                                                            },
                                                        ],
                                                    },
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 12,
                                                            col: 9,
                                                        },
                                                        end: SourcePos {
                                                            line: 12,
                                                            col: 29,
                                                        },
                                                    },
                                                },
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 10,
                                                    col: 9,
                                                },
                                                end: SourcePos {
                                                    line: 12,
                                                    col: 29,
                                                },
                                            },
                                        },
                                        right: AnnProc {
                                            proc: Send {
                                                channel: NameVar(
                                                    Id(
                                                        Id {
                                                            name: "world2",
                                                            pos: SourcePos {
                                                                line: 13,
                                                                col: 9,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                send_type: Single,
                                                inputs: [
                                                    AnnProc {
                                                        proc: StringLiteral(
                                                            "Hello World again",
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 13,
                                                                col: 17,
                                                            },
                                                            end: SourcePos {
                                                                line: 13,
                                                                col: 36,
                                                            },
                                                        },
                                                    },
                                                ],
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 13,
                                                    col: 9,
                                                },
                                                end: SourcePos {
                                                    line: 13,
                                                    col: 37,
                                                },
                                            },
                                        },
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 10,
                                            col: 9,
                                        },
                                        end: SourcePos {
                                            line: 13,
                                            col: 37,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 9,
                                    col: 5,
                                },
                                end: SourcePos {
                                    line: 14,
                                    col: 6,
                                },
                            },
                        },
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 4,
                            col: 5,
                        },
                        end: SourcePos {
                            line: 14,
                            col: 6,
                        },
                    },
                },
            },
            span: SourceSpan {
                start: SourcePos {
                    line: 3,
                    col: 1,
                },
                end: SourcePos {
                    line: 15,
                    col: 2,
                },
            },
        },
    ],
)
